### ISSUE TITLE
Update API as discussed in #53

### ISSUE THREAD

**Issue by davidlmobley** [MobleyLab/SolvationToolkit#60]
As discussed in #53 , we need to make a number of changes to the API to make it easier to avoid tripping ourselves up later and to make this more generally useful. 

***Change to a model where the constructor just initializes and then components are added piecewise as desired***

For example
```python
mixture = MixtureSystem()
mixture.addComponent(label='phenol', smiles='c1ccc(cc1)O', number=1)
mixture.addComponent(label='toluene', smiles='Cc1ccccc1', number=10)
mixture.addComponent(label='cyclohexane', smiles='C1CCCCC1', number=500)
mixture.build(prmtop_filename='mixture.prmtop', inpcrd_filename='mixture.inpcrd')
```

Note that (see below) `number` will be an optional argument, as will `label`. i.e., this is a more low-level API and a higher level is possible.

***Make the `build` option take a variety of arguments to only get out what file types we want***

See example above. The `build()` function would take a bunch of optional arguments for the file types desired to be saved. It should store .sdf and .mol2 files of all components anyway, and a .pdb of the final system, but then optional arguments should be used to control whether AMBER .prmtop/.crd files, GROMACS .top/.gro, etc. are saved.

***Make it so this can easily and transparently be used for pure solutions, mixtures, and infinite dilution setups***

For example, we want to be able to do something like this:
```python
liquid = MixtureSystem()
liquid.addComponent('water')

binary_mixture = MixtureSystem()
binary_mixture.addComponent(name='water', mole_fraction=0.2)
binary_mixture.addComponent(name='methanol') # assumed to be rest of mixture if no mole_fraction specified

ternary_mixture = MixtureSystem()
binary_mixture.addComponent(name='ethanol', mole_fraction=0.2)
binary_mixture.addComponent(name='methanol', mole_fraction=0.2)
ternary_mixture.addComponent(name='water')

infinite_dilution = MixtureSystem()
infinite_dilution.addComponent(name='phenol', mole_fraction=0.0) # infinite dilution
infinite_dilution.addComponent(name='water')
```

Note that this means:
- The `addComponent` will need to be able to accept a `name` argument (which would serve BOTH as the label AND as a source of canonical isomeric SMILES if these are not provided)
- Provision of `label` will be used for file names whether or not name is provided, i.e. `binary_mixture.addComponent(name='ethanol', label='test', mole_fraction=0.2)` means all ethanol files get called `test` and the name is only used in generating SMILES
- Provision of `name`, `label` AND `smiles` results in `label` and `smiles` being used and results in a warning that the name is not being used
- Given that this uses mole fraction, the code needs to make a sensible decision as to target box size which should be able to be overridden (i.e.  via the constructor) if the user desires
- If mole fractions of no components are provided (and not solute numbers), it will successfully set up the system only if a single component is provided (in which case it's a pure solution) and raise an exception if not (since solute numbers or mole fractions must be provided then) (get a recommendation from @bannanc as to what she's typically using as she's studied this).
- If mole fractions of all components are provided than it can set them up based on the default box size 
- In the infinite dilution case, it will need to recognize that a `mole_fraction` of 0.0 means only a single solute of each component with this mole fraction is desired. It should be acceptable to combine this with a mixture, i.e. you should be able to have components with mole fractions 0.4 and 0.6 that comprise the majority of the system, along with a single molecule of several components at mole fractions 0.0 and 0.0. 

***minor cleanup***
Strip out the `run` option from the class for now, including `self.equilibrate` and self.production` and `just_build` options.

*Please do open an WIP PR as soon as you begin working on this and commit to/push to it regularly so I can monitor.*

**Comment by nividic** [comment]
I'm starting working on this now

**Comment by bannanc** [comment]
I was discussing the best way to implement the mole fraction with Gaetano this afternoon. We are concerned about having default box sizes as then you also need a density for each molecule or an average density for the system in order to get a total number of molecules. We are thinking the most simple way to implement would be to have a default number of total molecules (maybe how many water molecules needed to make a large enough box for one of our simulations). Then there is no dependence on density.

**Comment by bannanc** [comment]
@davidlmobley and @nividic

**Comment by davidlmobley** [comment]
> I was discussing the best way to implement the mole fraction with Gaetano this afternoon. We are concerned about having default box sizes as then you also need a density for each molecule or an average density for the system in order to get a total number of molecules.

So, the thought is that one will always run a constant pressure equilibration on these systems which will bring things to the right density and pressure. 

One can't avoid the density issue, I don't think. Our current protocol uses a rough estimate of the density (currently, that of water) to get the box volume from the number of molecules specified. Unless we are willing to require the user to input a target number of molecules and target volume, there is no way around this, I believe. 

> We are thinking the most simple way to implement would be to have a default number of total molecules

I don't see how this avoids the issue; actually in my view it makes it worse, as if you have large molecules then having 1000 (for example) of them will result in very large large boxes whereas having 1000 small molecules will result in small boxes. 

In my view the best solution is to say something like, "Well, we know we typically need a box which is at least as large as X, so let's specify we want a box of that size, and then use a rough guess of the density to convert that to an approximate number of molecules." How to guess the density? Well, my approach was to use that of water, since it's towards the heavier end of common solvents (https://www.organicdivision.org/orig/organic_solvents.html) so if the actual density is lower than that of water you'll end up (after constant pressure equilibration) with a larger box than is really needed. That's a "safe" error to make, as problems only will arise when you end up with a smaller box than you wanted. 

That explains my rationale. But: 
> We are thinking the most simple way to implement would be to have a default number of total molecules (maybe how many water molecules needed to make a large enough box for one of our simulations). Then there is no dependence on density.

I still don't see how this would remove any density dependence, i.e. if you say you want 1000 cyclohexane molecules to simulate pure cyclohexane, you still have to figure out what size box you want to put them in, so you either need to know the density of cyclohexane, or just guess the density (i.e., that it has the same density as water) and let the equilibration sort it out. 

Or maybe there is some density dependence you're concerned about that I'm still missing?

**Comment by davidlmobley** [comment]
Resolved by #64
